pytorch_lightning/core/hooks.py

```diff
@@ -153,3 +155,48 @@ def backward(self, use_amp, loss, optimizer):
                     scaled_loss.backward()
         else:
             loss.backward()
+
+    def transfer_batch_to_device(self, batch: Any, device: torch.device) -> Any:
+        """
+        Override this hook if your :class:`~torch.utils.data.DataLoader` returns tensors
+        wrapped in a custom data structure.
+
+        The data types listed below (and any arbitrary nesting of them) are supported out of the box:
+
+        - :class:`torch.Tensor`
+        - :class:`list`
+        - :class:`dict`
+        - :class:`tuple`
+        - ``torchtext.data.Batch`` (COMING SOON)
+
+        For anything else, you need to define how the data is moved to the target device (CPU, GPU, TPU, ...).
+
+        Example::
+
+            def transfer_batch_to_device(self, batch, device)
+                if isinstance(batch, CustomBatch):
+                    # move all tensors in your custom data structure to the device
+                    batch.samples = batch.samples.to(device)
+                    batch.targets = batch.targets.to(device)
+                else:
+                    batch = super().transfer_batch_to_device(data, device)
+                return batch
+
+        Args:
+            batch: A batch of data that needs to be transferred to a new device.
+            device: The target device as defined in PyTorch.
+
+        Returns:
+            A reference to the data on the new device.
+
+        Note:
+            This hook should only transfer the data and not modify it, nor should it move the data to
+            any other device than the one passed in as argument (unless you know what you are doing).
+            The :class:`~pytorch_lightning.trainer.trainer.Trainer` already takes care of splitting the
+            batch and determines the target devices.
+
+        See Also:
+            - :func:`~pytorch_lightning.utilities.apply_func.transfer_batch_to_device`
+            - :func:`~pytorch_lightning.utilities.apply_func.apply_to_collection`
+        """
+        return transfer_batch_to_device(batch, device)
```

Comment:
This will be confusing having two function with the same name...
Make the lower call as protected

Reply:
you mean import `transfer_batch_to_device` as `_transfer_batch_to_device`?

Reply:
is this recursive?

Reply:
yes it is. it will descend into the nested structures like dicts etc. 
When the user overrides the hook, they need to call their hook recursively if their custom datastructure has a nested form, or they can call `super().transfer_batch_to_device` if their custom datastructure has dicts and lists that need to be transfered.

Reply:
I think that Will asked about the call if it is recursive, then no as this is a class method and called `transfer_batch_to_device` is a function from another package
then sorry also for the comment as there is no need for rename from my side :]